### DISCUSSION TITLE
[Feature Request]: Possibility to manually apply offset for temperature, humidity and barometric pressure measurements (BME280 and other sensors)

### DISCUSSION THREAD

**Discussion by Synthacon** [meshtastic/firmware#5482] [Ideas]
### Platform

ESP32

### Description

It would be useful to have built-in support for applying user-defined offsets to the temperature, humidity, and barometric pressure readings for the BME280 results. This sensor can't be recalibrated and humidity value drifts with the age. This option would allow users to easily adjust for systematic inaccuracies or environmental influences (e.g., heat from nearby components).

Correct for localized heating affecting temperature and humidity readings.
Adjust for known inaccuracies in factory calibration.
Enable quick fine-tuning for specific environmental conditions.

Just as example my sensor shows +0.64c and -10% for humidity and currently there is no way to adjust this.

**Comment by HarukiToreda** [comment]
this is indeed a problem plaguing most BME680 sensors as well, as they are not calibrated for accuracy at factory and there's no way to do so on the board itself. It must be offset by the firmware to be able to get accurate readings. All these sensors use temperature as a way to also calculate humidity and IAQ so having this wrong will lead to false readings all across the board rendering all inaccurate.

**Comment by h0lad** [comment]
They are measuring the PCB temperature and not environmental temperature. So maybe they are working as designed?

If there's a LDO next to them they might heat up the entire board which will influence the temperature measurement.

**Comment by Synthacon** [comment]
It is the components on the board that are heating the PCB. Even short wires would probably cause the PCB to be heated by the Meshtastic device.

I would imagine that a simple offset adjustment in the software could be a solution, but posting this in the software section only resulted in the request being locked the next day.

The sensor data is also read very frequently. One reading every 1-15 minutes would be sufficient. Would be awesome to be able to adjust this also... Could reducing the reading frequency help remove the PCB heating? This would not resolve the issue if the wires are short, causing heat from the Meshtastic device to dissipate through the wires into the PCB.

Ultimately, implementing a simple offset adjustment in the software would be the easiest way to give users control and allow them to correct the values. There is x amount of heat dissipating trought the wires+components off the sensor PCB. 

As now the whole idea of connecting sensors (tmp+humidity) is kind of pointless, nobody will ever get correct readings with the current options. There will be always huge error.

**Reply by fifieldt** [reply]
Not locked, just moved to discussion. I feel if the problem and fix can be better defined we might be able to do something.

**Comment by Synthacon** [comment]
![Screenshot_2024-12-01-09-04-17-24_a3b3f7ff182fbca1933b088238f0f0da](https://github.com/user-attachments/assets/eb50a2a0-d707-4705-8306-dd9a2abb983c)

Example about how this is solved in basic smart home Android app. It's just offset, not real calibration.

**Comment by bigdogdylan00** [comment]
I want to voice my support for this as well. A group of us are designing a bunch of solar nodes to act as weather stations, and are concerned that we will get useless data from these. As it is, my T1000-E gives temperatures that are easily 10⁰F above ambient, making the readings difficult to use at best

**Comment by weerdenburg** [comment]
I’ve just started with Meshtastic and bought four of the nice green Heltec V3 devices. My idea was to address multiple needs at once:
measuring temperature and humidity in several rooms, giving my children a small display where they can read those values, and at the same time providing them with their own Meshtastic device.

What I immediately noticed is that all of my BME280 sensors show about a **+2 °C offset**. Of course, I can compensate for this when I process the JSON data from all nodes in Node-RED, but that still leaves the **incorrect values on the local display** in each room.

So yes, I would really like to encourage adding **a simple sensor offset setting** for the sensors that can now be connected and configured. In my opinion, without this option, the sensors are almost useless for the end user of the device itself; they only become useful after external processing, where an offset is applied.

**Reply by h0lad** [reply]
Just for explaining the offset: You are measuring the PCB temperature. Not the environmental temperature.

This is the reason why are lot of temperature measurement modules cut PCB around the sensor to minimize this effect.

**Reply by weerdenburg** [reply]
Yes, I understand the external factors and PCB/layout effects that can influence the measurement.

But the practical issue remains: many people assume they can **simply connect** one of the **supported sensors** to a Meshtastic device, **enable a few settings** in the app, and then trust the displayed values. In reality _(at least 4 times in my setup)_, the readings have a consistent offset by default, and there’s currently no way to apply a simple calibration/offset on the device. That makes it hard to use the local display as a reliable “room readout” without doing external correction downstream.

**Comment by HarukiToreda** [comment]
I'll look into this to add into baseUI, the apps are handled by other devs so that may need a separate PR for their respective sides once I got this working

**Comment by HarukiToreda** [comment]
Something like this?


https://github.com/user-attachments/assets/310e3a6a-75f8-4e18-9ad4-f3f9806a1f72

**Reply by CodeSkunk521** [reply]
Oh my god! Yes please!! I have this exact same problem.

**Comment by HarukiToreda** [comment]
PR Created
https://github.com/meshtastic/firmware/pull/9682
